### PR TITLE
Associate library materials effect with meshes

### PR DESCRIPTION
**Pull request by ahcorde**
This PR is related with this issue https://github.com/ignitionrobotics/ign-common/issues/91#issuecomment-759632701

The material could be defined in Collada like this:

```xml
  <library_materials>
    <material id="material-cube">
      <instance_effect url="#effect-cube"/>
    </material>
  </library_materials>
```

This kind of material definition were never associated with the mesh. This PR try to fix this issue.

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Review by mjcarroll — APPROVED** [review]

**Comment by chapulina** [comment]
Would it be possible to add a test? Maybe using the mesh provided in #91 ?

**Comment by ahcorde** [comment]
+1 for the test. Added here https://github.com/ignitionrobotics/ign-common/pull/151/commits/188d06d0ee5d1b9ed3d3afcd2576e268606e8cec

File: graphics/src/ColladaLoader_TEST.cc [hunk 1]
```diff
@@ -51,6 +51,34 @@ TEST_F(ColladaLoader, LoadBox)
   EXPECT_STREQ("Cube", mesh->SubMeshByIndex(0).lock()->Name().c_str());
 }
 
+/////////////////////////////////////////////////
+TEST_F(ColladaLoader, LoadRectangle)
+{
+  common::ColladaLoader loader;
+  common::Mesh *mesh = loader.Load(
+      std::string(PROJECT_SOURCE_PATH) + "/test/data/rectangle.dae");
```
Comment (chapulina) [inline graphics/src/ColladaLoader_TEST.cc:59]: Let's use `common::joinPaths` for cross-platform support

File: graphics/src/ColladaLoader_TEST.cc [hunk 2]
```diff
@@ -51,6 +51,34 @@ TEST_F(ColladaLoader, LoadBox)
...
+  EXPECT_EQ(35u, mesh->VertexCount());
+  EXPECT_EQ(35u, mesh->NormalCount());
+  EXPECT_EQ(36u, mesh->IndexCount());
+  EXPECT_EQ(0u, mesh->TexCoordCount());
+  EXPECT_EQ(1u, mesh->SubMeshCount());
+  EXPECT_EQ(1u, mesh->MaterialCount());
+
+  common::MaterialPtr material = mesh->MaterialByIndex(0);
+  EXPECT_NE(nullptr, material);
+  EXPECT_EQ(math::Color(0.008957128, 0.00619848, 0.64, 1),
```
Comment (chapulina) [inline graphics/src/ColladaLoader_TEST.cc:74]: To fix Windows warning

```suggestion
  EXPECT_EQ(math::Color(0.008957128f, 0.00619848f, 0.64f, 1.0f),
```

The same for all colors below.

File: graphics/src/ColladaLoader_TEST.cc [hunk 3]
```diff
@@ -51,6 +51,34 @@ TEST_F(ColladaLoader, LoadBox)
...
+TEST_F(ColladaLoader, LoadRectangle)
+{
+  common::ColladaLoader loader;
+  common::Mesh *mesh = loader.Load(
+      std::string(PROJECT_SOURCE_PATH) + "/test/data/rectangle.dae");
+
+  EXPECT_STREQ("unknown", mesh->Name().c_str());
+  EXPECT_EQ(ignition::math::Vector3d(0.35, 0.4, 1), mesh->Max());
+  EXPECT_EQ(ignition::math::Vector3d(-0.35, -0.4, -1), mesh->Min());
+  // 36 vertices, 24 unique, 12 shared.
```
Comment (chapulina) [inline graphics/src/ColladaLoader_TEST.cc:64]: This comment doesn't seem right

File: graphics/src/ColladaLoader.cc [hunk 4]
```diff
@@ -549,6 +554,41 @@ void ColladaLoaderPrivate::LoadNode(tinyxml2::XMLElement *_elem, Mesh *_mesh,
...
+      iter = this->materialEffectMap.find(geomURL);
+      if (iter != this->materialEffectMap.end())
+        matStr = iter->second;
+      MaterialPtr mat = this->LoadMaterial(matStr);
+      matIndex = _mesh->IndexOfMaterial(mat.get());
+      if (matIndex < 0)
+      {
+        matIndex = _mesh->AddMaterial(mat);
+      }
+      if (matIndex < 0)
```
Comment (chapulina) [inline graphics/src/ColladaLoader.cc:574]: Shouldn't this block be combined with the one above?

Comment (ahcorde) [reply graphics/src/ColladaLoader.cc:574]: Not really, the first condional override the variable

https://github.com/ignitionrobotics/ign-common/blob/188d06d0ee5d1b9ed3d3afcd2576e268606e8cec/graphics/src/ColladaLoader.cc#L570-L577

Comment (chapulina) [reply graphics/src/ColladaLoader.cc:574]: ah yes!

**Comment by peci1** [comment]
I'm sorry, @ahcorde , but this PR seems to break more things than fix. X1 SubT robots are no longer blue, EXPLORER_DS1 has a weird red-ish color instead of carbon black...

image

Check the behavior of ColladaLoader with `~/.ignition/fuel/fuel.ignitionrobotics.org/openrobotics/models/x1 config 1/18/meshes/chassis.dae`.

This file has one `<vertices>` and two `<triangles>`. Each triangle list defines a different symbol for material. However, after loading the mesh with ColladaLoader, both submeshes have material index 1 assigned. Without this PR, submesh 0 correctly has material index 0. Here's a simple (wrongly) failing test for me:

```C++
TEST_F(ColladaLoader, LoadX1)
{
  common::ColladaLoader loader;
  common::Mesh *mesh = loader.Load("/home/peci1/.ignition/fuel/fuel.ignitionrobotics.org/openrobotics/models/x1 config 1/18/meshes/chassis.dae");

  ASSERT_EQ(2u, mesh->SubMeshCount());
  ASSERT_EQ(2u, mesh->MaterialCount());

  EXPECT_EQ(0u, mesh->SubMeshByIndex(0).lock()->MaterialIndex());
  EXPECT_EQ(1u, mesh->SubMeshByIndex(1).lock()->MaterialIndex());
}
```

Reading through #91 and Collada spec, I think the mesh submitted by @jaelrod is not what he expected it to be. The "triangles" section (p. 188) says:

> material   xs:NCName   Declares a symbol for a material. This symbol is bound to a material at the time of instantiation; see <instance_geometry> and <bind_material>. Optional. If not specified then the lighting and shading results are application defined. 

@jaelrod was concerned about reusability of the library geometries. But I think there was a misunderstanding on what does the `<triangles material="">` attribute actually mean. It does NOT specify the material to use. It specifies the symbol to which a material has to be bound in the `<instance_geometry>`. So it's not a "value", but a "variable name". I think there's nothing against naming the "variable" (material), as long as you can bind different materials to it (referencing the name in the `<instance_material symbol="">` attribute and specifying the selected material in `<instance_material target="">` attribute).

The mesh from #91 could work if the default value for the `<triangles material="">` were `"material"`. But the spec says that if this attribute is missing, shading is application-dependent. This is probably where application behavior starts to differ (and that's correct, as the spec says). To be compatible with Blender (which is generally a nice thing, I would say), this PR should be reverted and the change should be somewhere near

https://github.com/ignitionrobotics/ign-common/blob/abfecc8a262b32e3df7495f11b04fd38f5594f68/graphics/src/ColladaLoader.cc#L2009-L2029

I think this `if` could assume that there is a default value for the `material` attribute - `"material"`. It should just silence the warnings in case it tried to find a material and did not succeed.

**Pull request merged**